File: QS/quicksort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <chrono> // For timing

// Function to generate a vector filled with random integers
std::vector<int> generateRandomIntegers(size_t count) {
    std::vector<int> vec(count);
    std::srand(static_cast<unsigned>(std::time(nullptr))); // Seed for random number generation
    std::generate(vec.begin(), vec.end(), []() { return std::rand() % 10000; }); // Random numbers between 0 and 9999
    return vec;
}

// Functional style quicksort implementation
void quicksort(std::vector<int>& vec, int low, int high) {
    auto partition = [&](int low, int high) {
        int pivot = vec[high]; // Pivot
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (vec[j] < pivot) {
                i++;
                std::swap(vec[i], vec[j]); // Place all smaller elements to left of pivot
            }
        }
        std::swap(vec[i + 1], vec[high]);
        return i + 1;
    };

    if (low < high) {
        int pi = partition(low, high);
        quicksort(vec, low, pi - 1);  // Before partition index
        quicksort(vec, pi + 1, high); // After partition index
    }
}

// Function to find elements containing the digit '2'
void findElementsWithTwo(const std::vector<int>& vec) {
    for (int num : vec) {
        if (std::to_string(num).find('2') != std::string::npos) {
            std::cout << num << " ";
        }
    }
}

int main() {
    std::vector<int> numbers = generateRandomIntegers(10000); // Generate 10 000 random integers

    // Time the sorting process
    auto start = std::chrono::high_resolution_clock::now();
    quicksort(numbers, 0, numbers.size() - 1); // Apply quicksort
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::milli> sort_time = end - start;

    std::cout << "Sorting took " << sort_time.count() << " milliseconds.\n";
    std::cout << "After sorting:\n";
    for (int i = 0; i < 10; ++i) { // Print the first 10 elements
        std::cout << numbers[i] << " ";
    }
    std::cout << "...\n";

    // Time the search for elements containing the digit '2'
    start = std::chrono::high_resolution_clock::now();
    std::cout << "Elements containing '2': ";
    findElementsWithTwo(numbers);
    end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::milli> find_time = end - start;

    std::cout << "\nFinding elements took " << find_time.count() << " milliseconds.\n";

    return 0;
}

File: QS/quicksort.py
import random
import time

# Function to generate a list filled with random integers
def generate_random_integers(count):
    return [random.randint(0, 9999) for _ in range(count)]

# Quick sort implementation (in-place)
def quicksort(arr, low, high):
    def partition(low, high):
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        return i + 1
    
    if low < high:
        pi = partition(low, high)
        quicksort(arr, low, pi - 1)
        quicksort(arr, pi + 1, high)

# Function to find elements containing the digit '2'
def find_elements_with_two(arr):
    return [num for num in arr if '2' in str(num)]

def main():
    numbers = generate_random_integers(10000)  # Generate 10 000 random integers

    # Time the sorting process
    start_time = time.time()
    quicksort(numbers, 0, len(numbers) - 1)  # Apply quicksort
    end_time = time.time()
    print(f"Sorting took {end_time - start_time:.4f} seconds.")

    print("After sorting:")
    print(numbers[:10], "...")  # Print the first 10 elements

    # Time the search for elements containing the digit '2'
    start_time = time.time()
    result = find_elements_with_two(numbers)
    end_time = time.time()
    print("Elements containing '2':", result[:10], "...")  # Print first 10 matches for brevity
    print(f"Finding elements took {end_time - start_time:.4f} seconds.")

if __name__ == '__main__':
    main()
